### ISSUE TITLE
Plugin not working?

### ISSUE THREAD

**Issue by amandamcg** [CalebJohn/joplin-inline-todo#9]
Testing on Joplin 2.8.7 (prod, darwin) mac.

I have a note named `TODO list` with the following
`<!-- inline-todo-plugin -->`

I have a test note with a name and the content 
`- [ ] @TODO this is a todo`

I turned the refresh to 1 min for testing. 

Even after waiting a bit, the TODO list note never updates. I see these errors in the console: 
```
index.html:1 Uncaught (in promise) TypeError: Cannot read properties of null (reading 'body')
    at /Users/<directory>/.config/joplin-desktop/tmp/plugin_plugin.calebjohn.todo.js:1
    at Generator.next (<anonymous>)
    at s (/Users/<directory>/.config/joplin-desktop/tmp/plugin_plugin.calebjohn.todo.js:1)
index.html:1 Uncaught (in promise) Error: Not Found
    at defaultAction.ts:16
    at Generator.next (<anonymous>)
    at fulfilled (/Applications/Joplin.app/Contents/Resources/app.asar/node_modules/@joplin/lib/services/rest/utils/defaultAction.js:5)
    at runMicrotasks (<anonymous>)
    at processTicksAndRejections (internal/process/task_queues.js:95)
```

**Comment by amandamcg** [comment]
Note, this was my first install, I havent tried it before.

**Comment by CalebJohn** [comment]
It looks like the plugin is running in to an error and failing. It's interesting that I haven't seen it before. I'll push out a patch tomorrow that at the very least will prevent the crash, but will hopefully log some useful information to see why this is failing.

**Comment by CalebJohn** [comment]
I've pushed out a patch now that will fix the error that you saw. But I don't think that error would cause the summary note to not be generated.
The summary note is generated when you navigate from a different note into the summary note. Can you double check that it's not generating by navigating away from the `TODO list` and navigate back to it?

**Comment by franz-sp** [comment]
Hi CalebJohn, 
Thanks for this new great plugin!

Strangely enough I have a similar experience as amandamcg, but the plugin just shows one and the same todo, although I have many scattered in my notes.
That happens on both my Macs, a desktop MacPro and a MacBook Pro, both running MacOS 12.3.1.
My Joplin version is 2.7.15 (prod, darwin), Revision: 8352e23.
I tried to uninstall it and reinstall the plug in, nothing happens, except the line with exactly the same todo. 
I also tried to disable all the plug-ins except this one, but no joy.
I attach a log, in case it might be useful.

Thanks in advance

Francesco
[-1652978750746.log](https://github.com/CalebJohn/joplin-inline-todo/files/8732293/-1652978750746.log)
!

**Comment by CalebJohn** [comment]
> This apparent limitation could be actually intended as a feature: one can keep in the same note both "local" (plain) todos, as minor reminders or reusable checklists, that will not appear in the Joplin-wide todo list, and "global", application-wide, important todos (in confluence style), and keep an eye on them in the TODO list! With the added bonus that one can have an overview of "local" todos simply by changing the plug-in settings.

This is pretty close to how I use it. I make heavy use of regular `- [ ]` checkboxes, so the list style is almost useless for me. But I often will have tasks that use both, where the confluence style todo has the task description and I use checkboxes for sub-tasks. Once I've check off all the subtasks, I check the todo. Example:

```
- [ ] @todo Add table summary style
- [ ] Add new summary formatter
- [ ] Add details to readme
- [ ] prepare a new release
```

Setting "List Style" should show all checkboxes including confluence style todos. I'll look into this right away and hopefully come up with a fix.

**Comment by CalebJohn** [comment]
@franz-sp I was not able to duplicate the following issue

> If I write confluence-style todos, they correctly show when I select the "Confluence style" in my settings, but I can't see them if I switch back to the "List Style". Setting "Link Style" won't show any todo.

It works fine on my computer. Would it be possible for you to share an example of a todo that only works under the confluence style, but not as list style?

**Comment by franz-sp** [comment]
Oh my fault!

They actually appear in the list! 
I have just a small number of then, and probably missed them among the regular checkboxes, or it was just matter of time since the generation of the list is very slow (I have lots of check-lists made with checkboxes).
They appear now and then scattered along the whole list.

Sorry if I wasted your time on a non-existent issue :(

Thanks you again for your amazing work!

**Comment by CalebJohn** [comment]
@franz-sp no worries, I'm glad it's working for you.

**Comment by benmordecai** [comment]
I am getting the following errors

```
15:37:09: PluginService: Loading plugin from C:/Users/ben mordecai/.config/joplin-desktop/cache/plugin.calebjohn.todo
C:\Program Files\Jop…n\lib\Logger.js:190 15:37:10: joplin.plugins: Starting plugin: plugin.calebjohn.todo
C:\Program Files\Jop…n\lib\Logger.js:190 15:37:11: joplin.plugins: Finished running onStart handler: plugin.calebjohn.todo (Took 1019ms)
DevTools failed to load source map: Could not parse content for file:///C:/Program%20Files/Joplin/resources/app.asar/newTodo.js.map: Unexpected end of JSON input
```

**Comment by benmordecai** [comment]
Plugin is doing absolutely nothing.

**Comment by CalebJohn** [comment]
@benmordecai Do any other plugins work?
What version of Joplin are you using?

I can't reproduce this issue. I even tried using a username with a space in it (long shot chance).

**Comment by benmordecai** [comment]
All the other plugins work except the MacOS plugin which is crashing on multiple computers. I just tried the todo plugin on Arch linux and it similarly did not work. I'm away from the windows computer now but the linux version is 2.9.17

**Comment by CalebJohn** [comment]
How many notes do you have? Could be that the plugin is just slow

**Comment by benmordecai** [comment]
Roughly 600-700 notes

**Comment by sothawo** [comment]
Alas I'm having the same behaviour as @benmordecai 

When I write a todo I see in the marldown viewer that styling is applied to the assignee and tag, but the summary note stayz empty and does not show any entry.

Running Joplin 2.11.11 on macOS, Inline TODO version 1.6.0

**Comment by sothawo** [comment]
when switching to the summary note, the dev console shows
```
Uncaught (in promise) Error: Not Found
    at defaultAction.ts:16:21
    at Generator.next (<anonymous>)
    at fulfilled (/Applications/Joplin…faultAction.js:5:58)
    at process.processTicksAndRejections (node:internal/process/task_queues:96:5
```

**Comment by sothawo** [comment]
I downloaded and built the plugin and added some console logs.
Building the list gets stuck pn https://github.com/CalebJohn/joplin-inline-todo/blob/master/src/builder.ts#L27 

I changed the code to 
```
console.log("search in note " + note.title)
console.log(note)
let matches = [];
// This introduces a small risk of a race condition
// (If this is waiting, the note.body could become stale, but this function would
// continue anyways and update the summary with stale data)
// I don't think this will be an issue in practice, and if it does crop up
// there won't be any data loss
let folder = await this.get_parent_title(note.parent_id);
console.log("folder " + folder)
```
it gets stuck on getting the folder name fo that note

![grafik](https://github.com/CalebJohn/joplin-inline-todo/assets/8456476/910c9c24-f0c7-414d-833b-9428bc9ba8d5)

**Comment by CalebJohn** [comment]
@sothawo Thank you for doing all this reproduction. This is very helpful. Do you know if this note still exists, or if it's been deleted? I'm wondering if maybe there are some orphan notes lying around that don't technically have parents.

**Comment by CalebJohn** [comment]
@sothawo When you have a chance, can you please test [this branch](https://github.com/CalebJohn/joplin-inline-todo/tree/folder_race)? It adds early cancellation to the api call that the plugin is getting stuck on. If this solves the issue then I'll merge it into master (while trying to figure out why Joplin is stuck indefinitely in that call).

https://github.com/CalebJohn/joplin-inline-todo/tree/folder_race

**Comment by sothawo** [comment]
Thanks for the hint with the orphaned note. The note in question is one from the initial setup of Joplin, called "4. Tips". I have moved them to separate folders for the notes coming from the mobile and the desktop app as seen in the image.

![grafik](https://github.com/CalebJohn/joplin-inline-todo/assets/8456476/22021cda-c149-48af-93b5-55080f2ee985)

Yesterday I already had tried to move them without success. This morning I renamed both of them to "4. Tipss" with a double s ate the end and noticed that the console log still showed searching for the parent of "4. Tips". I searched for "4. Tips" then in Joplin and found another note with this name being orphaned.

I guess this might be a remnant of when I once split my notes into two profiles, but later rejoined them again in a single one as I don't like the process of restarting the app when switching profiles. I suppose in this export/import process this note might have been left.

I now created a notebook "orphaned", moved that file there an now all is working as it should. Alas I cannot move that back to have no parent now, so I cannot test your branch with that.

But I did a global search for "e" and was able to find other orphaned notes. I added a todo to one - after adding "orphaned" to it's name, and it got stuck on this very file.

I checked out your branch and added two more loggings:
```
			let unknown_folder = "Unknown Folder";
			let api_call = Promise.race([
				joplin.data.get(['folders', id], { fields: ['title'] }),
				new Promise((resolve, reject) => setTimeout(resolve, 1000, unknown_folder)),
			]);

			console.log("before api_call")
			let f = await api_call;
 			console.log("after api_call")
```
I never got to see the after api_call message, it gets stuck after the before api_call. I tried to play with the timeout, setting it to very short value, no success. btw, shouldn't the timeout function return `{title: unknown_folder}`?

Still trying things while writing this comment. I think I finally got it. The Joplin API throws an error not found (the one I posted above) which bubbles out stopping the whole search process. 

I'll post a PR shortly fixing this issue with getting the not existing parent.
